src/mercury_prof_pvar_impl.h

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (C) 2013-2020 Argonne National Laboratory, Department of Energy,
+ *                    UChicago Argonne, LLC and The HDF Group.
+ * All rights reserved.
+ *
+ * The full copyright notice, including terms governing use, modification,
+ * and redistribution, is contained in the COPYING file that can be
+ * found at the root of the source code distribution tree.
+ */
+
+#ifndef MERCURY_PROF_PVAR_IMPL_H
+#define MERCURY_PROF_PVAR_IMPL_H
+
+#include "mercury_prof_types.h"
+#include "mercury_hash_table.h"
+
+/*************************************/
+/* Public Type and Struct Definition */
+
+/*************************************/
+
+/* Internal PVAR data structure used to store PVAR data */
+struct hg_prof_pvar_data_t {
+   hg_prof_class_t pvar_class; /* PVAR class */
+   hg_prof_datatype_t pvar_datatype; /* PVAR datatype */
+   hg_prof_bind_t pvar_bind; /* PVAR binding */
+   int continuous; /* Is PVAR continuous or not */
+   void *addr; /* PVAR address */
+   int count; /* count of PVAR values */
+   char name[128]; /* PVAR name */
+   char description[128]; /* PVAR description */
+};
+
+typedef struct hg_prof_pvar_data_t hg_prof_pvar_data_t;
+
+extern hg_hash_table_t *pvar_table; /* Internal hash table containing all the PVAR information */
+
+/*****************/
+/* Public Macros */
+/*****************/
+
+#define NUM_PVARS 1 /* Number of PVARs currently exported. PVAR indices go from 0......(NUM_PVARS - 1). */
+
+/* PVAR declaration and registration macros */
+#define HG_PROF_PVAR_UINT_COUNTER_DECL(name) \
+    unsigned int PVAR_COUNTER_##name;
+
+#define HG_PROF_PVAR_UINT_COUNTER_DECL_EXTERN(name) \
+    extern unsigned int PVAR_COUNTER_##name;
+
+#define HG_PROF_PVAR_UINT_COUNTER_REGISTER(dtype, bind,\
+            name, desc) \
+        void *addr; \
+        /* Set initial value */ \
+        PVAR_COUNTER_##name = 0; \
+        addr = &PVAR_COUNTER_##name; \
+        HG_PROF_PVAR_REGISTER_impl(HG_PVAR_CLASS_COUNTER, dtype, #name, \
+            addr, 1, bind, 1, desc); 
+
+/* Increment the value of a PVAR */
+#define HG_PROF_PVAR_COUNTER_INC(name, val) \
+    *(&PVAR_COUNTER_##name) += val;
+
```

Comment:
@srini009 watching your video, it made me think that you'll need to make sure that all increments are either done using atomics or are protected by a mutex. I think it makes sense to go with atomic types (see https://github.com/mercury-hpc/mercury/blob/master/src/util/mercury_atomic.h) rather than using locks.

Reply:
Do the performance counters actually need to be atomic?  I had assumed not, but I could be missing something.

Reply:
If you're counting the number of forward calls for example and increment the counter, this has to be protected by a lock or be done atomically.

Reply:
I definitely see that it could be updated (or accessed) concurrently, just wasn't sure about if that caused danger here or not.  But I guess "+= val" is technically a read/modify/write so you could lose updates if it's not atomic.

I'm on board; we should use the hg atomics.